dd-sdk-android-gradle-plugin/src/main/kotlin/com/datadog/gradle/plugin/DdNdkSymbolFileUploadTask.kt

```diff
@@ -0,0 +1,191 @@
+            return emptySet()
+        }
+    }
+
+    companion object {
+        internal const val TASK_NAME = "ddUploadNdkSymbolFiles"
+        internal const val KEY_NDK_SYMBOL_FILE = "ndk_symbol_file"
+        internal const val TYPE_NDK_SYMBOL_FILE = "ndk_symbol_file"
+        internal const val ENCODING = "application/octet-stream"
+        internal val SUPPORTED_ARCHS = setOf("armeabi-v7a", "arm64-v8a", "x86", "x86_64")
+        internal const val MAX_DATADOG_CI_FILE_LOOKUP_LEVELS = 4
+
+        internal val LOGGER = Logging.getLogger("DdSymbolFileUploadTask")
+
+        @Suppress("StringLiteralDuplication")
+        private fun resolveDatadogRepositoryFile(target: Project): File {
+            val outputsDir = File(target.buildDir, "outputs")
+            val reportsDir = File(outputsDir, "reports")
+            val datadogDir = File(reportsDir, "datadog")
+            return File(datadogDir, "repository.json")
+        }
+
+        internal fun findDatadogCiFile(projectDir: File): File? {
+            var currentDir: File? = projectDir
+            var levelsUp = 0
+            while (currentDir != null && levelsUp < MAX_DATADOG_CI_FILE_LOOKUP_LEVELS) {
+                val datadogCiFile = File(currentDir, "datadog-ci.json")
+                if (datadogCiFile.exists()) {
+                    return datadogCiFile
+                }
+                currentDir = currentDir.parentFile
+                levelsUp++
+            }
+            return null
+        }
+
+        @Suppress("MagicNumber", "ReturnCount")
+        private fun isAgp7OrAbove(): Boolean {
+            val version = Version.ANDROID_GRADLE_PLUGIN_VERSION
+            val groups = version.split(".")
+            if (groups.size < 3) return false
+            val major = groups[0].toIntOrNull()
+            val minor = groups[1].toIntOrNull()
+            val patch = groups[2].substringBefore("-").toIntOrNull()
+            if (major == null || minor == null || patch == null) return false
+            return major >= 7 && minor >= 0 && patch >= 0
+        }
+
+        private fun fixNativeOutputPath(taskFolder: File): File {
+            return taskFolder.parentFile.parentFile.takeIf { it.parentFile.name == "cxx" }
+                ?: taskFolder.parentFile.parentFile.parentFile.takeIf { it.parentFile.name == "cxx" }
+                ?: taskFolder
+        }
+
+        private fun getSearchDir(buildTask: TaskProvider<ExternalNativeBuildTask>, propertyName: String, providerFactory: ProviderFactory): Provider<File> {
+            return buildTask.flatMap { task ->
+                val soFolder = ExternalNativeBuildTask::class.memberProperties.find { it.name == propertyName }?.get(task)
+                when (soFolder) {
+                    is File -> providerFactory.provider { fixNativeOutputPath(soFolder) }
+                    is DirectoryProperty -> soFolder.map { fixNativeOutputPath(it.asFile) }
+                    else -> providerFactory.provider { File("") }
```

Comment:
I think changing method signature to `Provider<File?>` and returning `providerFactory.provider { null }` will be cleaner here.

Or maybe it is better even to throw and exception, we don't have anything to upload anyway, and it is during the task execution stage, not task configuration stage, so for the end-user it will be fine to see task failure (and we will know that we need a change here).

Reply:
I thought I couldn't change the return to `Provider<File?>` for some reason... 😅

I think throwing the exception is probably not the best solution because we can have multiple .so files, and I don't think we want to fail all the others if just one fails.

Reply:
I do get an issue returning null occasionally complaining that the property has no value, but in functional testing this has gone away now that everything surrounding it is working....

Reply:
Maybe then use `Provider#isPresent` before querying the value if using `Provider<File?>`. My concern here is that if in some future version type of `soFolder` is not `File` or `DirectoryProperty`, then we will silently miss it by using hacky `File("")` return value. We should notify about that.

Reply:
I changed it to return null, which will fail properly.